src/controllers/profileController.ts

```diff
@@ -24,12 +27,47 @@ import { profileMissingErrorResponse } from "../types/profile";
 import { withUserFromRequest } from "../types/user";
 import { withValidatedOrValidationError } from "../utils/responses";
 
-export default class ProfileController {
+import { IBackendController } from "./IBackendController";
+
+export default class ProfileController implements IBackendController {
   constructor(
     private readonly profileService: ProfileService,
     private readonly sessionStorage: ISessionStorage
   ) {}
 
+  // tslint:disable-next-line: no-any
+  public setupRouting<ResBody = any, ReqBody = any>(
+    app: express.Express,
+    basePath: string,
+    ...handlers: ReadonlyArray<
+      RequestHandler<ParamsDictionary, ResBody, ReqBody>
+    >
+  ): void {
+    app.get(
```

Comment:
A drawback of this approach is that now the paths are "hidden" inside each controller and it harder to catch overlapping paths

Reply:
Yes, as I listed in "Cons", this could happen.
But I think that over time the _app.ts_ file could grow by an average of 10 lines per each new api function exposed, and this, in my opinion, could lead similar problems.

Reply:
A smart move to overcome this problem could be to use `express.Router`:
```ts
const routerA = express.Router();
routerA.get("/foo", fooA); 
const routerB = express.Router();
routerB.get("/foo", fooB); 
const app = express();
app.use("/a", routerA);
app.use("/b", routerB);
// GET /a/foo --> fooA
// GET /b/foo --> fooB
```
We can then consider a `Controller` as an application that takes one (or more) `Service` and returns an `express.Router`:
```ts
const myController = (myService) => {
  const router = express.router();
  /* 
  * bind middlewares and handlers to the routes
  */
  return router;
}
```
And, in `app.ts`, we can bind whole routers to their sub paths; this way we can keep every sub path accessible at a glance in one single place.

Other benefit: authentication middleware can be also handled outside the controller, at router level. Example:
```ts
app = express();

app.use(
  PagoPABasePath,
  checkIP(allowPagoPAIPSourceRange),
  bearerWalletTokenAuth,
  pagopaRouter
);

app.use(
  registerMyPortalRoutes,
  checkIP(allowMyPortalIPSourceRange),
  authMiddlewares.bearerMyPortal
  myPortalRouter
);
```

(Please note that this is no much different from the actual usage of `register<XYZ>Routes` helper functions, but maybe it can enable more modularity).

Reply:
Even better, so Controllers don't even need to know about auth middlewares.

Reply:
@balanza  I tried to move auth middleware outside the controller, but I got test failure from ""Test the checkIP middleware" > "should allow in-range IP"" test (we get 400 instead of 200).

Reply:
AFAIK middleware execution is not affected by where you put them (but I might be wrong).

Just be careful that when you apply a middleware to a router, it's used by all its endpoint

Reply:
That's what I thought, and I checked that all endpoints needed the middleware in place, I don't understand..